File: go/98.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// use left bound and right bound to test
func helper(t *TreeNode, l, r int) bool {
	if t == nil {
		return true
	}
	if t.Val > r || t.Val < l {
		return false
	}
	return helper(t.Left, l, t.Val) && helper(t.Right, t.Val, r)
}

func isValidBST(root *TreeNode) bool {
	return helper(root, -1<<32, 1<<32)
}

File: python/98.py
from typing import Optional
# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def helper(self, root: Optional[TreeNode], lb: int, rb: int):
        if root is None:
            return True
        elif root.left and root.val <= root.left.val:
            return False
        elif root.right and root.val >= root.right.val:
            return False
        elif root.val <= lb or root.val >= rb:
            return False
        return self.helper(root.left, lb, root.val) and \
            self.helper(root.right, root.val, rb)

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.helper(root, -2**31-1, 2**31)
